File: lib/controllers/store_controller.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:frontend/pages/store/store_list_page.dart';
import 'package:http/http.dart';
import '../models/store.dart';
import '../utils/strings.dart';

class StoreController {
  Future<List<Store>> listStore() async {
    try {
      Response res = await get(
        Uri.parse('$BASE_URL$STORE_URL'),
        headers: {AUTH: '1'},
      );

      if (res.statusCode == 200) {
        List<dynamic> body = jsonDecode(res.body);

        List<Store> store =
            body.map((dynamic item) => Store.fromJson(item)).toList();

        return store;
      } else {
        throw "Não foi possível carregar as lojas";
      }
    } catch (e) {
      print('Error ' + e);
      return null;
    }
  }

  Future<void> saveStore(String name, String description, context) async {
    const Utf8Codec utf8 = Utf8Codec();

    try {
      Response res = await post(
        Uri.parse('$BASE_URL$STORE_URL'),
        encoding: utf8,
        body: jsonEncode(
          {'name': name, 'description': description},
        ),
        headers: {
          AUTH: '1',
          'Content-Type': 'application/json',
        },
      );

      if (res.statusCode == 200) {
        print("POST");
        Navigator.push(
          context,
          MaterialPageRoute(
            builder: (context) => StoreListPage(),
          ),
        );
      } else {
        throw "Não foi possível cadastrar a loja";
      }
    } catch (e) {
      print('Error ' + e);
    }
  }

  Future<void> deleteStore(int id, context) async {
    try {
      Response res = await delete(
        Uri.parse("$BASE_URL$STORE_URL/$id"),
        headers: {AUTH: '$id'},
      );

      if (res.statusCode == 200) {
        print("DELETED");
        Navigator.push(
          context,
          MaterialPageRoute(
            builder: (context) => StoreListPage(),
          ),
        );
      } else {
        throw "Não foi possível deletar a loja";
      }
    } catch (e) {
      print('Error ' + e);
    }
  }
}

File: src/controllers/StoreController.js
const connection = require('../database/connection');

module.exports = {
  async get(req, res) {
    const {id} = req.query;
    const authorization = req.headers.authorization;

    if (!id) {
      const storeList = await connection('store')
        .select('*');

      return res.json(storeList);

    } else {
      const storeListId = await connection('store')
        .select('*')
        .where('id', id);

      return res.json(storeListId);
    } 
  },

  async post(req, res) {
    const {name, description} = req.body;

    const authorization = req.headers.authorization;

    await connection('store')
      .insert({
        name,
        description
      });

    return res.json({success: true});
  },

  async put(req, res) {
    const {id} = req.params;
    const {name, description} = req.body;
    const authorization = req.headers.authorization;

    await connection('store')
      .where('id', id)
      .update({
        name,
        description
      })
      .then(rows => {
        if (!rows) {
          return res.status(404).json({success: false});
        }
        
        return res.json({success: true});
      })
      .catch(error => {
        return res.status(500).json({success: false});
      })
  },

  async patch(req, res) {
    const {id} = req.params;
    const {name, description} = req.body;
    const authorization = req.headers.authorization;

    await connection('store')
      .where('id', id)
      .update({
        name,
        description
      })
      .then(rows => {
        if (!rows) {
          return res.status(404).json({success: false});
        }
        
        return res.json({success: true});
      })
      .catch(error => {
        return res.status(500).json({success: false});
      })
  },

  async delete(req, res) {
    const {id} = req.params;
    const authorization = req.headers.authorization;

    const storeDeleteId = await connection('store')
      .where('id', id)
      .select('id')
      .first();

    if (parseInt(storeDeleteId.id) !== parseInt(authorization)) {
      return res.status(401).json({success: false});
    }

    try {
      await connection('store')
      .where('id', id)
      .delete();

      return res.status(204).json({success: true});

    } catch(error) {
      console.log(error);

      return res.status(401).json({success: false});
    }
  }
}
